File: backend/internal/handlers/checklist.go
package handlers

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/go-chi/chi/v5"
	"realtourflow/internal/middleware"
)

type checklistItem struct {
	ID         string  `json:"id"`
	DealID     string  `json:"deal_id"`
	Label      string  `json:"label"`
	Category   string  `json:"category"`
	Checked    bool    `json:"checked"`
	AssignedTo string  `json:"assigned_to"`
	DueDate    *string `json:"due_date,omitempty"`
	IsCustom   bool    `json:"is_custom"`
	SortOrder  int     `json:"sort_order"`
}

type defaultItem struct {
	Label      string
	Category   string
	AssignedTo string
}

var defaultChecklistItems = []defaultItem{
	{Label: "Contract received and reviewed", Category: "Contract", AssignedTo: "tc"},
	{Label: "Earnest money deposit verified", Category: "Contract", AssignedTo: "tc"},
	{Label: "All parties have signed contract", Category: "Contract", AssignedTo: "tc"},
	{Label: "Loan application submitted", Category: "Loan", AssignedTo: "tc"},
	{Label: "Disclosures out", Category: "Loan", AssignedTo: "tc"},
	{Label: "Disclosures signed and submitted", Category: "Loan", AssignedTo: "tc"},
	{Label: "Approved with conditions", Category: "Loan", AssignedTo: "tc"},
	{Label: "Appraisal ordered", Category: "Loan", AssignedTo: "tc"},
	{Label: "Clear to close received", Category: "Loan", AssignedTo: "tc"},
	{Label: "Title ordered", Category: "Title", AssignedTo: "tc"},
	{Label: "Title search complete", Category: "Title", AssignedTo: "tc"},
	{Label: "Title commitment received", Category: "Title", AssignedTo: "tc"},
	{Label: "Wire instructions confirmed", Category: "Title", AssignedTo: "tc"},
	{Label: "Closing date confirmed with all parties", Category: "Closing", AssignedTo: "tc"},
	{Label: "Closing disclosure sent", Category: "Closing", AssignedTo: "tc"},
	{Label: "Final walkthrough scheduled", Category: "Closing", AssignedTo: "agent"},
	{Label: "Keys and access items prepared", Category: "Closing", AssignedTo: "tc"},
}

var checklistEligibleStages = map[string]bool{
	"under_contract": true,
	"pre_close":      true,
	"closing":        true,
	"post_close":     true,
}

func (h *Handler) checklistAccess(r *http.Request, dealID, userID string) bool {
	for _, role := range middleware.GetRoles(r) {
		if role == "tc" || role == "admin" {
			return true
		}
	}
	var count int
	h.db.QueryRowContext(r.Context(), `
		SELECT COUNT(*) FROM deals
		WHERE id = $1 AND (
			agent_id = $2 OR
			EXISTS (SELECT 1 FROM deal_participants WHERE deal_id = $1 AND user_id = $2)
		)
	`, dealID, userID).Scan(&count)
	return count > 0
}

// ListChecklist returns checklist items for a deal.
// Auto-seeds defaults if the deal is in an eligible stage and has no items yet.
func (h *Handler) ListChecklist(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r)
	if claims == nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	userID, err := resolveUserID(r.Context(), h.db, claims.RegisteredClaims.Subject)
	if err != nil {
		http.Error(w, "user not found", http.StatusNotFound)
		return
	}
	dealID := chi.URLParam(r, "dealId")
	if !h.checklistAccess(r, dealID, userID) {
		http.Error(w, "deal not found", http.StatusNotFound)
		return
	}

	// Check if items already exist
	var count int
	h.db.QueryRowContext(r.Context(), `SELECT COUNT(*) FROM checklist_items WHERE deal_id = $1`, dealID).Scan(&count)

	// Auto-seed if eligible stage and empty
	if count == 0 {
		var stage string
		h.db.QueryRowContext(r.Context(), `SELECT stage FROM deals WHERE id = $1`, dealID).Scan(&stage)
		if checklistEligibleStages[stage] {
			for i, item := range defaultChecklistItems {
				h.db.ExecContext(r.Context(), `
					INSERT INTO checklist_items (deal_id, label, category, assigned_to, sort_order)
					VALUES ($1, $2, $3, $4, $5)
				`, dealID, item.Label, item.Category, item.AssignedTo, i)
			}
		}
	}

	rows, err := h.db.QueryContext(r.Context(), `
		SELECT id, deal_id, label, category, checked, assigned_to, due_date, is_custom, sort_order
		FROM checklist_items
		WHERE deal_id = $1
		ORDER BY sort_order, created_at
	`, dealID)
	if err != nil {
		http.Error(w, "failed to fetch checklist", http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	items := make([]checklistItem, 0)
	for rows.Next() {
		var it checklistItem
		var dueDateRaw *time.Time
		if err := rows.Scan(&it.ID, &it.DealID, &it.Label, &it.Category, &it.Checked, &it.AssignedTo, &dueDateRaw, &it.IsCustom, &it.SortOrder); err != nil {
			http.Error(w, "failed to scan item", http.StatusInternalServerError)
			return
		}
		if dueDateRaw != nil {
			s := dueDateRaw.Format("2006-01-02")
			it.DueDate = &s
		}
		items = append(items, it)
	}

	respond(w, http.StatusOK, items)
}

// CreateChecklistItem adds a custom item to a deal's checklist.
func (h *Handler) CreateChecklistItem(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r)
	if claims == nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	userID, err := resolveUserID(r.Context(), h.db, claims.RegisteredClaims.Subject)
	if err != nil {
		http.Error(w, "user not found", http.StatusNotFound)
		return
	}
	dealID := chi.URLParam(r, "dealId")
	if !h.checklistAccess(r, dealID, userID) {
		http.Error(w, "deal not found", http.StatusNotFound)
		return
	}

	var req struct {
		Label      string `json:"label"`
		Category   string `json:"category"`
		AssignedTo string `json:"assigned_to"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || req.Label == "" {
		http.Error(w, "label is required", http.StatusBadRequest)
		return
	}
	if req.Category == "" {
		req.Category = "Contract"
	}
	if req.AssignedTo == "" {
		req.AssignedTo = "tc"
	}

	var maxOrder int
	h.db.QueryRowContext(r.Context(), `SELECT COALESCE(MAX(sort_order), -1) FROM checklist_items WHERE deal_id = $1`, dealID).Scan(&maxOrder)

	var it checklistItem
	if err := h.db.QueryRowContext(r.Context(), `
		INSERT INTO checklist_items (deal_id, label, category, assigned_to, is_custom, sort_order)
		VALUES ($1, $2, $3, $4, TRUE, $5)
		RETURNING id, deal_id, label, category, checked, assigned_to, due_date, is_custom, sort_order
	`, dealID, req.Label, req.Category, req.AssignedTo, maxOrder+1).Scan(
		&it.ID, &it.DealID, &it.Label, &it.Category, &it.Checked, &it.AssignedTo, &it.DueDate, &it.IsCustom, &it.SortOrder,
	); err != nil {
		http.Error(w, "failed to create item", http.StatusInternalServerError)
		return
	}
	respond(w, http.StatusOK, it)
}

// UpdateChecklistItem patches checked, assigned_to, or due_date on a checklist item.
func (h *Handler) UpdateChecklistItem(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r)
	if claims == nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	userID, err := resolveUserID(r.Context(), h.db, claims.RegisteredClaims.Subject)
	if err != nil {
		http.Error(w, "user not found", http.StatusNotFound)
		return
	}
	dealID := chi.URLParam(r, "dealId")
	itemID := chi.URLParam(r, "itemId")
	if !h.checklistAccess(r, dealID, userID) {
		http.Error(w, "deal not found", http.StatusNotFound)
		return
	}

	var req struct {
		Checked    *bool   `json:"checked"`
		AssignedTo *string `json:"assigned_to"`
		DueDate    *string `json:"due_date"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "invalid body", http.StatusBadRequest)
		return
	}

	if req.Checked != nil {
		h.db.ExecContext(r.Context(), `UPDATE checklist_items SET checked = $1, updated_at = NOW() WHERE id = $2 AND deal_id = $3`, *req.Checked, itemID, dealID)
	}
	if req.AssignedTo != nil {
		h.db.ExecContext(r.Context(), `UPDATE checklist_items SET assigned_to = $1, updated_at = NOW() WHERE id = $2 AND deal_id = $3`, *req.AssignedTo, itemID, dealID)
	}
	if req.DueDate != nil {
		if *req.DueDate == "" {
			h.db.ExecContext(r.Context(), `UPDATE checklist_items SET due_date = NULL, updated_at = NOW() WHERE id = $1 AND deal_id = $2`, itemID, dealID)
		} else {
			h.db.ExecContext(r.Context(), `UPDATE checklist_items SET due_date = $1, updated_at = NOW() WHERE id = $2 AND deal_id = $3`, *req.DueDate, itemID, dealID)
		}
	}

	var it checklistItem
	var dueDateRaw *time.Time
	if err := h.db.QueryRowContext(r.Context(), `
		SELECT id, deal_id, label, category, checked, assigned_to, due_date, is_custom, sort_order
		FROM checklist_items WHERE id = $1 AND deal_id = $2
	`, itemID, dealID).Scan(
		&it.ID, &it.DealID, &it.Label, &it.Category, &it.Checked, &it.AssignedTo, &dueDateRaw, &it.IsCustom, &it.SortOrder,
	); err != nil {
		http.Error(w, "item not found", http.StatusNotFound)
		return
	}
	if dueDateRaw != nil {
		s := dueDateRaw.Format("2006-01-02")
		it.DueDate = &s
	}
	respond(w, http.StatusOK, it)
}

// DeleteChecklistItem removes a checklist item (custom items only for non-TC roles).
func (h *Handler) DeleteChecklistItem(w http.ResponseWriter, r *http.Request) {
	claims := middleware.GetClaims(r)
	if claims == nil {
		http.Error(w, "unauthorized", http.StatusUnauthorized)
		return
	}
	userID, err := resolveUserID(r.Context(), h.db, claims.RegisteredClaims.Subject)
	if err != nil {
		http.Error(w, "user not found", http.StatusNotFound)
		return
	}
	dealID := chi.URLParam(r, "dealId")
	itemID := chi.URLParam(r, "itemId")
	if !h.checklistAccess(r, dealID, userID) {
		http.Error(w, "deal not found", http.StatusNotFound)
		return
	}

	h.db.ExecContext(r.Context(), `DELETE FROM checklist_items WHERE id = $1 AND deal_id = $2`, itemID, dealID)
	respond(w, http.StatusOK, map[string]string{"status": "ok"})
}

File: web/lib/checklist.ts
import { prisma } from "./db";
import { isLinkedTCForDeal } from "./deals";
import { hasRole } from "./roles";

export const CHECKLIST_ELIGIBLE_STAGES = new Set([
  "under_contract",
  "pre_close",
  "closing",
  "post_close",
]);

export type DefaultChecklistItem = {
  label: string;
  category: string;
  assignedTo: "tc" | "agent" | "buyer" | "seller" | "third_party";
};

// ── Seller-portal defaults (#261) ────────────────────────────────────────────
// The seller portal's Listing Prep and Pre-Close cards are backed by real,
// persisted checklist_items (assigned_to='seller') instead of cosmetic local
// state. These sets are keyed by deal TYPE + STAGE (see sellerDefaultsFor), NOT
// by CHECKLIST_ELIGIBLE_STAGES, so they seed at active_search / pre_close for
// sell deals only. Labels intentionally match the seller-portal card copy so a
// seeded item shows up as the same row the seller already expects to see.

// Seeded at active_search ("Listing Prep") for sell deals.
export const SELLER_LISTING_PREP_ITEMS: DefaultChecklistItem[] = [
  { label: "Deep clean / declutter", category: "Listing Prep", assignedTo: "seller" },
  { label: "Minor repairs completed", category: "Listing Prep", assignedTo: "seller" },
  { label: "Professional photos scheduled", category: "Listing Prep", assignedTo: "seller" },
  { label: "Listing copy approved", category: "Listing Prep", assignedTo: "seller" },
  { label: "Disclosures package complete", category: "Listing Prep", assignedTo: "seller" },
  { label: "Lockbox installed", category: "Listing Prep", assignedTo: "seller" },
];

// Seeded at pre_close for sell deals.
export const SELLER_PRE_CLOSE_ITEMS: DefaultChecklistItem[] = [
  { label: "Complete agreed repairs", category: "Pre-Close", assignedTo: "seller" },
  { label: "Remove all personal belongings", category: "Pre-Close", assignedTo: "seller" },
  { label: "Schedule final walkthrough access", category: "Pre-Close", assignedTo: "seller" },
  { label: "Confirm possession date", category: "Pre-Close", assignedTo: "seller" },
  { label: "Utilities transfer arranged", category: "Pre-Close", assignedTo: "seller" },
];

// Which seller default set (if any) belongs to a deal at a given type + stage.
// Only sell deals seed seller items; buy deals get none here (their loan-side
// coordination lives in the TC set below). Returns [] when nothing applies.
export function sellerDefaultsFor(
  type: string,
  stage: string
): DefaultChecklistItem[] {
  if (type !== "sell") return [];
  if (stage === "active_search") return SELLER_LISTING_PREP_ITEMS;
  if (stage === "pre_close") return SELLER_PRE_CLOSE_ITEMS;
  return [];
}

export const DEFAULT_CHECKLIST_ITEMS: DefaultChecklistItem[] = [
  { label: "Contract received and reviewed", category: "Contract", assignedTo: "tc" },
  { label: "Earnest money deposit verified", category: "Contract", assignedTo: "tc" },
  { label: "All parties have signed contract", category: "Contract", assignedTo: "tc" },
  { label: "Loan application submitted", category: "Loan", assignedTo: "tc" },
  { label: "Disclosures out", category: "Loan", assignedTo: "tc" },
  { label: "Disclosures signed and submitted", category: "Loan", assignedTo: "tc" },
  { label: "Approved with conditions", category: "Loan", assignedTo: "tc" },
  { label: "Appraisal ordered", category: "Loan", assignedTo: "tc" },
  { label: "Clear to close received", category: "Loan", assignedTo: "tc" },
  { label: "Title ordered", category: "Title", assignedTo: "tc" },
  { label: "Title search complete", category: "Title", assignedTo: "tc" },
  { label: "Title commitment received", category: "Title", assignedTo: "tc" },
  { label: "Wire instructions confirmed", category: "Title", assignedTo: "tc" },
  { label: "Closing date confirmed with all parties", category: "Closing", assignedTo: "tc" },
  { label: "Closing disclosure sent", category: "Closing", assignedTo: "tc" },
  { label: "Final walkthrough scheduled", category: "Closing", assignedTo: "agent" },
  { label: "Keys and access items prepared", category: "Closing", assignedTo: "tc" },
];

export async function checklistHasAccess(
  dealId: string,
  userId: string,
  roles: readonly string[]
): Promise<boolean> {
  if (hasRole(roles, ["admin"])) return true;
  // A TC only has access when the deal's owning agent linked them (#172).
  if (hasRole(roles, ["tc"]) && (await isLinkedTCForDeal(dealId, userId))) {
    return true;
  }
  const deal = await prisma.deals.findFirst({
    where: {
      id: dealId,
      OR: [
        { agent_id: userId },
        { deal_participants: { some: { user_id: userId } } },
      ],
    },
    select: { id: true },
  });
  return !!deal;
}
